File: Java/deepspace/src/main/java/deepspace/SpaceStation.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import static java.lang.Float.max;
import static java.lang.Float.min;
import java.util.ArrayList;

/**
 *
 * @author fjsae
 */
public class SpaceStation {
    private static final float MAXFUEL = 100f;
    private static final float SHIELDLOSSPERUNITSHOT =0.1f;
    
    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private float shieldPower;
    
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;
    private Hangar hangar; // this will be o or 1
    private Damage pendingDamage; // this will be 0 or 1
    
    public String toString(){
        String ret = String.format("This SpaceStation has (ammoPower,fuelUnits,name,nMedals,shieldPower)=(%f,%f,%s,%d,%f)",ammoPower,fuelUnits,name,nMedals,shieldPower);
        
        String weap = "\n+ WEAPONS MOUNTED: \n";
        if(weapons == null || weapons.isEmpty())
            weap += " Ninguna\n";
        else{
            for(Weapon w : weapons){
                weap += w.toString();
            }
        }
        
        String shB = "\n+ SHIELDBOOSTERS MOUNTED: \n";
        if(shieldBoosters == null || shieldBoosters.isEmpty())
            shB += " Ninguno\n";
        else{
            for(ShieldBooster sB : shieldBoosters){
                shB += sB.toString();
            }
        }
        
        String han = "\n* HANGAR: \n";
        if(hangar == null)
            han += " Ninguno\n";
        else
            han += hangar.toString();
        
         
        String pDam = "\n* PENDINGDAMAGE: \n";
        if(pendingDamage == null)
            pDam += " Ninguno\n";
        else
            pDam += pendingDamage.toString() + "\n";
                
        return ret + weap + shB + han + pDam;
                
    }
    
    private void assignFuelValue(float f){
        this.fuelUnits = min(f,MAXFUEL);
    }
    
    private void cleanPendingDamage(){
        if(this.pendingDamage.hasNoEffect()){
            this.pendingDamage = null;
        }
    }
    
    SpaceStation(String n, SuppliesPackage supplies){
        this.name = n;
        this.ammoPower = supplies.getAmmoPower();
        assignFuelValue(supplies.getFuelUnits());
        this.shieldPower = supplies.getShieldPower();
        this.weapons = new ArrayList<Weapon>();
        this.shieldBoosters = new ArrayList<ShieldBooster>();
        this.pendingDamage = null;
        this.hangar = null;
        
    }

    public float getAmmoPower() {
        return ammoPower;
    }

    public float getFuelUnits() {
        return fuelUnits;
    }

    public String getName() {
        return name;
    }

    public int getNMedals() {
        return nMedals;
    }

    public float getShieldPower() {
        return shieldPower;
    }

    public ArrayList<Weapon> getWeapons() {
        return weapons;
    }

    public ArrayList<ShieldBooster> getShieldBoosters() {
        return shieldBoosters;
    }

    public Hangar getHangar() {
        return hangar;
    }

    public Damage getPendingDamage() {
        return pendingDamage;
    }
    
    public SpaceStationToUI getUIversion(){
        return new SpaceStationToUI(this);
    }
    
    public float getSpeed(){
        return this.fuelUnits/MAXFUEL;
    }
    
    public void cleanUpMountedItems(){
        for(int i = 0; i < this.weapons.size(); i++){
            if(this.weapons.get(i).getUses() == 0){
                this.weapons.remove(i);
                i--;
            }
                
        }
        for(int i =0; i < this.shieldBoosters.size(); i++){
            if(this.shieldBoosters.get(i).getUses() == 0){
                this.shieldBoosters.remove(i);
                i--;
            }
        }
    }
    
    public void discardHangar(){
        this.hangar = null;
    }
    
    public void discardshieldBooster(int i){
        if(i >= 0 && i < this.shieldBoosters.size()){
            ShieldBooster s  = this.shieldBoosters.remove(i);
            if(this.pendingDamage != null){
                this.pendingDamage.discardShieldBooster();
                cleanPendingDamage();
            }
        }
    }
    
    public void discardShieldBoosterInHangar(int i){
        if(this.hangar != null){
            this.hangar.removeShieldBooster(i);
        }
    }
    
    public void discardWeapon(int i){
        if(i >= 0 && i < this.weapons.size()){
            Weapon w  = this.weapons.remove(i);
            if(this.pendingDamage != null){
                this.pendingDamage.discardWeapon(w);
                cleanPendingDamage();
            }
        }
    }
    
    
    public void discardWeaponInHangar(int i){
        if(this.hangar != null){
            this.hangar.removeWeapon(i);
        }
    }
    
    public float fire(){
        float factor = 1.0f;
        for(Weapon w: this.weapons){
            factor *= w.useIt();
        }
        return factor*this.ammoPower;
    }
    
    public float protection(){
        float factor = 1.0f;
        for(ShieldBooster s: this.shieldBoosters){
            factor *= s.useIt();
        }
        return factor*this.shieldPower;
    }
    
    
    public void mountShieldBooster(int i){
        ShieldBooster b = this.hangar.removeShieldBooster(i);
        if(b != null){
            this.shieldBoosters.add(b);
        }
    }
    
    public void mountWeapon(int i){
        Weapon w = this.hangar.removeWeapon(i);
        if(w != null){
            this.weapons.add(w);
        }
    }
    
    public void move(){
        this.fuelUnits = Float.max(this.fuelUnits - this.getSpeed(),0f);
    }
    

    public void receiveHangar(Hangar h){
        if(this.hangar == null){
            this.hangar = h;
        }
    }
    
    public boolean receiveShieldBooster(ShieldBooster s){
        boolean ret = false;
        if(this.hangar!= null){
            ret = hangar.addShieldBoosters(s);
        }
        return ret;
    }
    
    public ShotResult receiveShot(float shot){
        ShotResult ret = ShotResult.DONOTRESIST;
        if (this.protection() >= shot){
            this.shieldPower -= max(0.0f,this.SHIELDLOSSPERUNITSHOT*shot);
            ret = ShotResult.RESIST;
        }   
        else{
            this.shieldPower = 0.0f;
        }
        return ret;
    }
    
    public void receiveSupplies(SuppliesPackage s){
        this.ammoPower += s.getAmmoPower();
        this.shieldPower += s.getShieldPower();
        assignFuelValue(this.fuelUnits + s.getFuelUnits());
        
    }
    
    public boolean receiveWeapon(Weapon w){
        boolean ret = false;
        if(this.hangar!= null){
            ret = hangar.addWeapon(w);
        }
        return ret;
    }
    
    public void setLoot(Loot loot){
        CardDealer dealer = CardDealer.getInstance();
        
        if(loot.getNHangars() > 0){
            Hangar hangar = dealer.nextHangar();
            this.receiveHangar(hangar);
        }
        
        int elements = loot.getNSupplies();
        for(int i = 0; i < elements; i++){
            SuppliesPackage sup = dealer.nextSuppliesPackage();
            this.receiveSupplies(sup);
        }
        
        elements = loot.getNWeapons();
        for(int i = 0; i < elements; i++){
            Weapon weap = dealer.nextWeapon();
            this.receiveWeapon(weap);
        }
        
        elements = loot.getNShields();
        for(int i = 0; i < elements; i++){
            ShieldBooster sh = dealer.nextShieldBooster();
            this.receiveShieldBooster(sh);
        }
        
        this.nMedals += loot.getNMedals();
        
            
    }
    
    public void setPendingDamage(Damage d){
        Damage adjusted = d.adjust(this.weapons,this.shieldBoosters);
        if(adjusted.getNWeapons() == 0 && adjusted.getNShields() == 0){
            adjusted = null;
        }
        this.pendingDamage = adjusted;
    }
    
    public boolean validState(){
        boolean ret = true;
        if(this.pendingDamage != null){
            if(!this.pendingDamage.hasNoEffect()){
                ret = false;
            }
        }
        return ret;
    }
    
    
}

File: Ruby/SpaceStation.rb
#encoding:utf-8

require_relative 'SpaceStationToUI'
require_relative 'SuppliesPackage'
require_relative 'CardDealer'
require_relative 'Damage'

module Deepspace

    class SpaceStation

        @@SHIELDLOSSPERUNITSHOT = 0.1
        @@MAXFUEL = 100.0
        

        def initialize(n,supplies)
            @name = n
            @ammoPower = supplies.ammoPower
            assignFuelValue(supplies.fuelUnits)
            @nMedals = 0
            @shieldPower = supplies.shieldPower
            @weapons = Array.new
            @shieldBoosters = Array.new
            @hangar = nil
            @pendingDamage = nil
        end

        def copy(station)
            @name = station.name
            @ammoPower = station.ammoPower
            assignFuelValue(station.fuelUnits)
            @nMedals = station.nMedals
            @shieldPower = station.shieldPower
            @weapons = station.weapons
            @shieldBoosters = station.shieldBoosters
            @hangar = station.hangar
            @pendingDamage = station.pendingDamage
        end

        def to_s
            namE = "+ NAME: #{@name}\n"
            aP = "+ AMMOPOWER: #{@ammoPower} "
            fU = "+ FUELUNITS: #{@fuelUnits} "
            sP = "+ SHIELDPOWER: #{@shieldPower}\n"
            nMed = "+ NMEDALS: #{@nMedals}\n"
            
            pDam = "+ PENDING DAMAGE: Ninguno \n"
            if pendingDamage != nil
              pDam = "+ PENDING DAMAGE: \n#{@pendingDamage.to_s}"
            end
            
            wP = "+ WEAPONS MOUNTED: Ninguno \n"
            if !weapons.empty?
              wP = "\n+ WEAPONS MOUNTED: \n"
              @weapons.each { |w|
                wP += "#{w.to_s}\n"
              }
            end
            
            sB = "+ SHIELDBOOSTERS MOUNTED: Ninguno \n"
            if !shieldBoosters.empty?
              sB = "\n+ SHIELDBOOSTERS MOUNTED: \n"
              @shieldBoosters.each { |shieldB|
                sB += "#{shieldB.to_s}\n"
              }
            end
            
            hG = "+ HANGAR: Ninguno \n"
            if hangar != nil
              hG = "\n+ HANGAR: \n#{@hangar.to_s}\n"
            end
            
            return namE + aP + fU + nMed + sP + pDam + wP + sB + hG
          end

        def cleanUpMountedItems
            auxWeapons = Array.new
            auxShields = Array.new

            @weapons.each do |w|
                if w.uses > 0
                    auxWeapons << w
                end
            end

            @shieldBoosters.each do |s|
                if s.uses > 0
                    auxShields << s
                end
            end

            @weapons = auxWeapons
            @shieldBoosters = auxShields
        end


        def assignFuelValue(f) 
            @fuelUnits = [f,@@MAXFUEL].min
        end
    
        def cleanPendingDamage 
            if @pendingDamage != nil
                if @pendingDamage.hasNoEffect
                    @pendingDamage = nil
                end
            end
        end

        def discardHangar
            @hangar = nil
        end

        def discardShieldBooster(i)
            if (i >= 0 && i < @shieldBoosters.length)
                s = @shieldBoosters.delete_at(i)

                if @pendingDamage != nil
                    @pendingDamage.discardShieldBooster
                    cleanPendingDamage()
                end
            end

        end

        def discardShieldBoosterInHangar(i)
            if @hangar != nil
                @hangar.removeShieldBooster(i)
            end
        end

        def discardWeapon(i)
            if (i >= 0 && i < @weapons.length)
                w = @weapons.delete_at(i)

                if @pendingDamage != nil
                    @pendingDamage.discardWeapon(w)
                    cleanPendingDamage()
                end
            end
        end

        def discardWeaponInHangar(i)
            if @hangar != nil
                @hangar.removeWeapon(i)
            end
        end

        def fire
            factor = 1
            @weapons.each do |w|
                factor *= w.useIt
            end
            factor = @ammoPower * factor
            factor
        end

        def speed
            @fuelUnits/@@MAXFUEL
        end

        def getUIversion
            SpaceStationToUI.new(self)
        end

        def mountShieldBooster(i)
            if @hangar != nil
                s = @hangar.removeShieldBooster(i)
                if s != nil
                    @shieldBoosters << s
                end
            end
        end

        def mountWeapon(i)

            if @hangar != nil
                w = @hangar.removeWeapon(i)
                if w != nil
                    @weapons << w
                end
            end

        end

        def move
            @fuelUnits = [@fuelUnits - speed(),0 ].max
        end

        def protection
            factor = 1
            @shieldBoosters.each do |s|
                factor *= s.useIt
            end
            factor = @shieldPower * factor
            factor
        end

        def receiveHangar(h)
            if @hangar == nil
                @hangar = h
            end

        end

        def receiveShieldBooster(s)
            ret = false
            if @hangar != nil
                ret = @hangar.addShieldBooster(s)
            end
            ret
        end

        def receiveShot(shot)
            myProtection = protection()
            ret = ShotResult::DONOTRESIST
            #puts "Me disparan con #{shot} y me protejo con #{myProtection}"
            if myProtection >= shot
                @shieldPower = [@shieldPower -@@SHIELDLOSSPERUNITSHOT*shot,0 ].max
                ret = ShotResult::RESIST
            else
                @shieldPower = 0.0

            end
            ret
        end

        def receiveSupplies(s)
            @ammoPower += s.ammoPower
            assignFuelValue(@fuelUnits + s.fuelUnits)
            @shieldPower += s.shieldPower

        end

        def receiveWeapon(w)
            ret = false 
            if @hangar != nil
                ret = @hangar.addWeapon(w)
            end
            ret
        end

        def validState
            ret = true
            if @pendingDamage != nil
                if !@pendingDamage.hasNoEffect
                    ret = false
                end
            end
            ret
        end

        def setPendingDamage(d)
            adj = d.adjust(@weapons,@shieldBoosters)
            @pendingDamage = adj
        end

        def setLoot(loot)

            dealer = CardDealer.instance

            if loot.nHangars > 0
                #puts "Mira como recibo un hangar"
                h = dealer.nextHangar
                #puts h.inspect
                receiveHangar(h)
            end

            elements = loot.nSupplies
            for i in 0..elements
                sup = dealer.nextSuppliesPackage
                receiveSupplies(sup)
            end

            elements = loot.nWeapons
            for i in 0..elements
                weap = dealer.nextWeapon
                receiveWeapon(weap)
            end

            elements = loot.nShields
            for i in 0..elements
                sh = dealer.nextShieldBooster
                receiveShieldBooster(sh)
            end

            @nMedals += loot.nMedals

            ret = Transformation::NOTRANSFORM
            if(loot.getEfficient)
                ret = Transformation::GETEFFICIENT
            elsif(loot.spaceCity)
                ret = Transformation::SPACECITY
            end

            ret
        end
        


        private :assignFuelValue, :cleanPendingDamage
        attr_reader :ammoPower, :fuelUnits, :hangar, :name, :nMedals, :pendingDamage, :shieldBoosters, :shieldPower, :weapons
        attr_writer :damage , :loot

    end

end